### DISCUSSION TITLE
AIR native build: Could not apply timestamp to setup.msi

### DISCUSSION THREAD

**Discussion by vvvdennis** [airsdk/Adobe-Runtime-Support#1193] [General]
Hello fellow Air fans & Hartman.

I am having this error trying to build native installer for extended desktop-enabled Air application on Windows using latest AIR SDK 33.1.1.575
I am doing 
**adt -package -target native -keystore mycert.pfx -storetype pkcs12 -tsa none MyAppInstal.exe MyAirFile.air**

I tried using -tsa with symantec time server, tried building directly from swf and source files. Same error. Around January 1st this year, this was working with AIR SDK 32, which I tried as well.

Can somebody help me?

**Comment by ajwfrost** [comment]
Hi

This error appears to be coming from the "nais.exe" tool from the SDK - but this isn't something we've changed! So I'm wondering if you could just copy the earlier file from an old SDK into this one (lib/nai/bin folder) and see if it helps?

We'll do a bit of debugging and try to find out why it's not picking up the timestamp properly..

thanks

**Reply by vvvdennis** [reply]
Thank you Andrew. 

I already did verify that using previous SDK throws same error. That would be something on Windows side behaving more strict than before. I have a theory that the certificate.pfx that I am using, it has "code signing" as its stated purpose, but not "time stamping" and maybe that is what they started to enforce somehow. That is just a theory though. It would be nice to understand what the inner error is actually about. 

I appreciate your help. I hope AIR is alive and well. On MAC, I am using application bundle with captive runtime to distribute this. Is this still the recommended way or should users download air.dmg? Note, application is using extended desktop capabilities.

**Reply by ajwfrost** [reply]
Can you please try with "-tsa http://timestamp.sectigo.com" and see if this changes things?

thanks

**Reply by ajwfrost** [reply]
Sorry just missed your second question -> yes, definitely best to create an application bundle with captive runtime. For new macOS versions it's also necessary to sign these with Developer ID Certificates, if you're not distributing via the App Store. For this you can use the `adt -sign` option with `-target bundle` and a store type `KeychainStore` with the alias being the name of your certificate...

**Reply by vvvdennis** [reply]
I tried it and now I got a different error:

Certificate in ..pathtomyfile.pfx could not be used to sign temp.exe

**Reply by vvvdennis** [reply]
Do you know what it means, Andrew? The certificate has not expired, but only has "code signing" as its purpose. This is a step forward, thank you!

**Reply by ajwfrost** [reply]
Just investigating this now.

I have found out that the timestamping server that was hard-coded into the tool originally by Adobe, is no longer in existence, so we'll update it to point to the replacement. But "could not be used to sign" can be output in one of three cases; interestingly when I run nais.exe on its own I don't see that, but when I run it from ADT then I see the same thing...

**Reply by ajwfrost** [reply]
Translating the call that's failing, so we can reproduce it using the Microsoft 'signtool.exe' tool, it's giving the same error:
```
SignTool Error: SignedCode::Sign returned error: 0x800700C1
        For more information, please see https://aka.ms/badexeformat
SignTool Error: An error occurred while attempting to sign: MyAppInstal.exe
```
The temp.exe file is showing up as a `PE32 executable (GUI) Intel 80386, for MS Windows` file, and is opening fine when I run it directly and when I open it as a zip file (it's a self-extracting zip).

So, not sure why it's considered to be invalid. The above link doesn't go anywhere useful. We'll have to look at this self-extracting exe mechanism..!

**Reply by ajwfrost** [reply]
Ah ... so now I can see why -> the self-extracting archive starts with the "sea.exe" file but we have code-signed this! and this signature is then preventing anything else being applied... so if we remove the signature from this file (`codesign remove /s sea.exe`) and try it again, it works fine.

Can you try this -> below is a zipped-up sea.exe file that needs to go into the SDK under lib/nai/lib.

thanks
[sea.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7237928/sea.zip)

**Reply by vvvdennis** [reply]
It resolved the error. 
Thanks much, Andrew.
